src/Microsoft.OpenApi/Models/OpenApiComponents.cs

```diff
@@ -154,33 +168,37 @@ public void SerializeAsV3(IOpenApiWriter writer)
             SerializeInternal(writer, OpenApiSpecVersion.OpenApi3_0, (writer, element) => element.SerializeAsV3(writer),
                 (writer, referenceElement) => referenceElement.SerializeAsV3WithoutReference(writer));
         }
-        
+
         /// <summary>
         /// Serialize <see cref="OpenApiComponents"/>.
         /// </summary>
-        private void SerializeInternal(IOpenApiWriter writer, OpenApiSpecVersion version, 
+        private void SerializeInternal(IOpenApiWriter writer, OpenApiSpecVersion version,
             Action<IOpenApiWriter, IOpenApiSerializable> callback, Action<IOpenApiWriter, IOpenApiReferenceable> action)
         {
             // Serialize each referenceable object as full object without reference if the reference in the object points to itself.
             // If the reference exists but points to other objects, the object is serialized to just that reference.
 
             // schemas
-            writer.WriteOptionalMap(
-                OpenApiConstants.Schemas,
-                Schemas,
-                (w, key, component) =>
+            if (Schemas31 != null && Schemas31.Any())
+            {
+                if (writer is OpenApiYamlWriter)
```

Comment:
I think we need to update the OpenApiJsonWriter to handle writing JsonSchema objects rather than lifting the code into the model objects.